pkg/test/framework/components/echo/options.go

```diff
@@ -0,0 +1,35 @@
+//  Copyright Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package echo
+
+import (
+	"fmt"
+)
+
+// NameForCluster will set a cluster-specific service name for each instance of the
+// config by appending '-{clusterIndex}' to the end of it's current service name.
+func NameForCluster() ConfigModifier {
+	return func(cfg *Config) {
+		cfg.Service += fmt.Sprintf("-%d", cfg.Cluster.Index())
+	}
+}
+
+func SetRef(refs ...*Instance) ConfigModifier {
```

Comment:
I'd like to keep SetRef for a few reasons:
- It would be very repetitious to have a `Get` call for each service in single cluster tests with many echos.
- More complex/manual refactoring, moving away from a well-established pattern. 
- We could actually use it cleanly for multicluster. One improvement could be having a SetRefs(*[]Instance) that will resize a slice according to the # of clusters in the ClusterContext. 

```go
    clients := make([]*echo.Instance, len(ctx.Clusters())
    servers := make([]*echo.Instance, len(ctx.Clusters())
    echoboot.NewBuilderOrFail(t).ForClusters().
      With(cfg, echo.SetRef(clients...)).
      With(cfg, echo.SetRef(servers...)).
      Done().BuildOrFail(t)
```

Reply:
Sure that's fine ... I'm fine leaving the existing interface for now to support single cluster cleanly.

For multicluster, do you have a good feel for what our tests need to do? For example, do they need to get all instances of a given app across all clusters? Or just one instance in some cluster?

Now that I think about it, Echo already supports multiple workloads in a single `Instance`. Perhaps we extend it to have workloads across multiple clusters? The workload could indicate which cluster it's installed on.  That would basically put us back to our current interface:

```golang
var a,b echo.Instance
echoboot.NewBuilderOrFail(t).
  With(&a, echo.Config {
    ...
    Service: "a",
    Clusters: ctx.Clusters(),
  }).
  With(&b, echo.Config {
    ...
    Service: "b",
    Clusters: ctx.Clusters(),
  }).
  BuildOrFail(t)
```

No `Ref` necessary. The only limitation to this will be that it will likely create the same number of workloads per cluster. We could add an option later to allow the user to configure per-cluster replicas.

WDYT?

Reply:
Makes sense, but I think it's honestly easier to write tests with separate echo instances way. Close to half of cases, maybe the majority, we want to make the services unique per-cluster with similar configs. 

The main advantage of this approach is it makes the refactor for building smaller, but at the same time we use echo.Instance rather than workloads to say "the instance in this cluster should send traffic to X".

Reply:
> we want to make the services unique per-cluster with similar configs.

Just to clarify, do you mean that in most cases, we want an app to only reside in a single cluster (e.g. "a" is in cluster-0 and "b" is in cluster-1)? Or do you mean that you may want an app deployed to multiple clusters but with slightly different configurations?

If the former, my concern would be that a more common configuration would be to spread services across all/most clusters.

If the latter, what sort of things would we be changing per-cluster?

Reply:
It would be a common config with a different service name. We could still do something like the ConfigModifier but on a per-workload basis and that'd do the same thing. 

An example:
https://github.com/istio/istio/pull/24549/files#diff-3c4dfbc0c4bc721eed6a8b0a6245cc0d (protocol_sniffing_test)

Reply:
I guess another quality of life thing with the ClusterContext approach is avoiding the need to pipe `clusters` into the many helper funcs for building cluster-configs

Reply:
Ah I see ... I feel like the [helper method approach](https://github.com/istio/istio/blob/master/tests/integration/multicluster/helper.go#L63) is reasonable here. I'm not sure I see this being a major quality of life problem ... in fact, duplicating the config inline seems worse IMO.

If we absolutely want to eliminate the need for per-test helper functions, we could add builder functions to the config to simplify common tasks:

```golang
cfg := &echo.Config {
...
}

var a,b echo.Instance
echoboot.NewBuilderOrFail(t).
  With(&a, cfg.WithService("a")).
  With(&b, cfg.WithService("b")).
  BuildOrFail(t)
```

WRT interacting with the same service in multiple clusters, we could also extract the `Call` methods into a base `Caller` interface and have it shared with `Workload`.  Then we could make it easy to call from specific clusters/workloads with something like:

```golang
a.WorkloadOrFail(ctx, cluster0).Call(opts)
```

This is actually similar to what is currently being done in a few places. Workload already has a `ForwardEcho` function which is ultimately what's called by `Call`. A few tests are calling this method directly. Adding the `Call` methods to workload probably makes sense in general.

WDYT?

Reply:
I don't think the doing workloads per-cluster will help for a lot of the tests. It's fine for traffic-senders, but we won't have a unique service name to send traffic to and we'd still end up needing multiple echo instances.

Reply:
That is unless we want to test these scenarios with default cross-cluster loadbalancing for a single service and validate that we hit all clusters using the ParsedResponse.

Reply:
Yeah, different versions would absolutely warrant different instances of echo. Basically what I was suggesting was to have a single `Instance` represent all of the echo workloads with the exact same configuration.

Reply:
Gotcha. So that's with keeping the ClusterContext part, bringing us here:

```go
client echo.Instance
echos := echoboot.NewBuilderOrFail(t).
  With(&client, echo.Config {
    ...
    Service: "client",
    Clusters: ctx.Clusters(),
  }).
  ForClusters().
    With(echo.Config {
      ...
      Service: "server",
    }, echo.NameForCluster()).
  Done().
  BuildOrFail(t)

for _, c := range ctx.Clusters() {
  target := echos.GetOrFail(ctx, echo.ServicePrefix("server").And(echo.InCluster(c)))
  client.WorkloadOrFail(ctx, c).Call(...)
}
```

Reply:
Ah right ... I think you had mentioned that use case before. If we have to create unique server instances in each cluster, how bad would it be to just have to do it manually? ...

```golang
// Create an echo client in each cluster.
client echo.Instance
b := echoboot.NewBuilderOrFail(t).
  With(&client, echo.Config {
    ...
    Service: "client",
    Clusters: ctx.Clusters(),
  })

// Add a uniquely named echo server instance to each cluster.
servers := make([]echo.Instance, len(ctx.Clusters()))
for i := range servers {
  b.With(&server[i], 
    echo.Config {
    ...
    Service: fmt.Sprintf("echo-server-%d", i),
    Cluster: ctx.Clusters()[i],
  }
}
 
_ = b.BuildOrFail(t)

// Send traffic from each cluster to every other cluster.
for _, src := range client.Workloads() {
  for _, dst := range servers {
    src.CallOrFail(ctx, /*options with dst*/)
  }
}
```

This way we keep the builder as simple as possible, but the Call interface on the Workload allows us to choose exactly where we're calling from.

Reply:
That's what I had before I dug into echo changes, 
It can be a little [cumbersome](https://github.com/istio/istio/pull/24549/commits/d9006d42876e03a2b784e6b86e54818b21c49e76#diff-3c4dfbc0c4bc721eed6a8b0a6245cc0dR79) but most of the pain can be dealt with by the `Instances` and matcher way of getting results.

Reply:
Yeah I think some of the additions (workloads can `Call`, and `Instances` util methods) should make it a little cleaner. Also getting rid of the pilot dependency will be a great cleanup.

Reply:
Agreed - scoped this down to just the Instances bit and usages in the existing multicluster tests. Splitting out removing pilot and workload callers to their own PRs.